### DISCUSSION TITLE
Support java.sql.Array and java.sql.Struct

### DISCUSSION THREAD

**Discussion by lukaseder** [duckdb/duckdb#7173] [Ideas]
I'm aware of these issues:

- An improvement allowing to fetch a `String` representation of nested data types: https://github.com/duckdb/duckdb/pull/7000
- A duplicate of this request (?) that I think was closed prematurely: https://github.com/duckdb/duckdb/issues/6559

While https://github.com/duckdb/duckdb/pull/7000 is already very helpful for tooling like Dbeaver, I wonder if there are still any plans of supporting composite types more formally via JDBC `java.sql.Array` and `java.sql.Struct`, despite #6559 having been closed. I think that a string representation of composite types won't be as good as a more formal representation.

To illustrate, there is (I think?) still no way to distinguish these very distinct arrays:

```sql
select 
  cast(['a','b'] as text) as a,
  cast(['a, b'] as text) as b
```

Producing:

```
|a     |b     |
|------|------|
|[a, b]|[a, b]|
```

Both resulting strings are identical. It doesn't matter if the strings are produced by SQL or the JDBC driver.

If the client is in control of the SQL, then it could wrap all projected columns in auxiliary function calls:

```sql
select 
  to_json(['a','b']) as a,
  to_json(['a, b']) as b
```

Now, we're getting:

```
|a        |b       |
|---------|--------|
|["a","b"]|["a, b"]|
```

But it would be much nicer if the JDBC driver offered formal support, both to read and write these data types.

Note that even Dbeaver, which does indeed implement some cleverness decoding the strings can't distinguish between the above:

<img width="322" alt="image" src="https://user-images.githubusercontent.com/734593/234077723-5716dc3d-ef97-4c80-a68f-15ddde0bf6da.png">

**Comment by Mause** [comment]
Being able to read structs and arrays is obviously required, but how important is being able to write them?

**Reply by lukaseder** [reply]
It's certainly less important:

1. They can be decomposed into their components, i.e. rather than binding just `?`, users can bind e.g. `ARRAY[?, ?]`
2. One of the biggest benefits of these types is projecting nested data structures directly from queries, much more than actually storing the types

So, I think that support for writing these types can definitely be delayed.

**Reply by Mause** [reply]
@lukaseder support for native `getArray` has been merged and will be part of the release. We're in the feature freeze now though, so I'm not sure I'll be able to get the Struct and Map support merged

**Reply by lukaseder** [reply]
Thanks, @Mause
